platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,709 @@
+is "bad". Or if the events it sends cannot be parsed, or are signed incorrectly, or in other ways fail validation, then
+it is "bad". There may be additional rules by the Gossip module or others (such as Event Intake detecting branching)
+that could lead to a neighbor being marked as "bad". A "bad" node may be dishonest, or it may be broken. The two cases
+may be indistinguishable, so punishment must be adjusted based on the severity of the behavior.
+
+If the Sheriff decides that the neighbor should be penalized, then it will instruct the Gossip module to "shun" that
+neighbor. "Shunning" is a unilateral behavior that one node can take towards another, where it terminates the connection
+and refuses to work further with that neighbor. If the Sheriff decides to welcome a neighbor back into the fold, it can
+instruct the Gossip module to "welcome" the neighbor back.
+
+#### Falling Behind
+
+When a node is operating, it receives events from its neighbors through gossip. If the node for some reason is unable to
+receive and process events fast enough, it may start to "fall behind" the other nodes. Perhaps Bob is processing round
+200 while Alice is still on round 100. If Alice continues to fall farther and farther behind, the time may come when she
+can no longer get old events from her neighbors. From the perspective of the neighbor, the events Alice says she needs
+may have expired, and Bob may no longer be holding those events.
+
+If this happens, then Alice has "fallen behind" and must reconnect. There is no longer any chance that she can get the
+events she needs through gossip alone. Gossip will detect this situation and make a call through the Consensus module
+interface to notify Execution that the node is behind. Execution will then initiate reconnect.
+
+Fundamentally, Execution is responsible for reconnect, but it cannot differentiate between a quiescent network or a
+behind node just by looking at the lack of rounds coming from Consensus. So Consensus **must** tell Execution if it
+knows that it cannot proceed in creating rounds because the node itself is bad, vs. a quiescent network which is
+able to create rounds (if there were any transactions).
+
+#### Roster Changes
+
+At runtime, it is possible that the roster will change dynamically (as happens with the dynamic address book feature).
+Roster changes at the gossip level may influence which neighbors the module will work with. As with all other modules
+using rosters, Gossip may need a deterministic understanding of which roster applies to which round. It will receive
+this information from Hashgraph in the form of round metadata.
+
+### Event Intake
+
+The Event Intake System is responsible for receiving events, validating them, and emitting them in *topological order*.
+It also makes sure they are durably persisted before emission so as to add resilience to the node in case of certain
+catastrophic failure scenarios.
```

Comment:
Yes, that's why I was very careful to refer to that as `a state` in my comment, and not as `the state`. And my point still stands - the Consensus has a state of its own, and hence it doesn't (currently) look simply as "a library" that is fed all the data it needs from elsewhere (e.g. from the Execution part.)

We need to decide if we're good with that, and if so, update the wording in the document accordingly.

If, on the other hand, we aren't good with that, then we need to modify the interfaces between the Consensus and the Execution to let the latter feed that "state" (not "the state") to the Consensus (and also let the Consensus update that "state" through an API on the Execution part.)

Reply:
> We need to decide if we're good with that, and if so, update the wording in the document accordingly.

Making the execution layer responsible for all on-disk data from the consensus/intake subsystems (e.g. the PCES) would result in some really ugly interactions between the two systems. I don't think any any sane design would result the execution layer being responsible for all on-disk data for consensus/intake. Additionally, if consensus and execution end up getting run on different computers, then such a split would introduce extreme technical challenges.

Reply:
I don't have a strong opinion one way or another, but I just want the document to be precise so that it's actually useful. Stating that Consensus is just a library that doesn't manage any state directly and then doing some I/O and persisting some data on disk is at the very least confusing.

Reply:
> But somewhere above in this document it is stated that Consensus doesn't have any state of its own.

I did my best to clarify the parts of the document which talk about the sort of state that consensus doesn't have.

Reply:
Thanks.